lib/api/core/httpRouter/index.js

```diff
@@ -135,99 +135,105 @@ class Router {
-      request.response.setHeaders(this.defaultHeaders);
+      routeHandler.request.response.setHeaders(this.defaultHeaders);
 
       if (routeHandler.handler === null) {
         errorsManager.throw('api_url_not_found', routeHandler.url);
       }
 
-      if (httpRequest.content.length <= 0) {
-        routeHandler.invokeHandler(cb);
-        return;
-      }
-
-      if (httpRequest.headers['content-type']
-      && !httpRequest.headers['content-type'].startsWith('application/json')) {
-        errorsManager.throw(
-          'invalid_request_content_type',
-          httpRequest.headers['content-type']);
-      }
-
-      {
-        const encoding = CharsetRegex.exec(httpRequest.headers['content-type']);
-
-        if (encoding !== null && encoding[1].toLowerCase() !== 'utf-8') {
-          errorsManager.throw(
-            'invalid_request_charset',
-            encoding[1].toLowerCase());
-        }
+      if (!message.isEmpty()) {
+        this.validateMessage(message);
       }
 
-      routeHandler.addContent(httpRequest.content);
       routeHandler.invokeHandler(cb);
-
-      return null;
     } catch (err) {
-      if (request === undefined) {
-        request = new Request({requestId: httpRequest.requestId}, {}, 'rest');
+      let request;
+      if (!routeHandler || !routeHandler._request) {
+        request = new Request({requestId: message.requestId}, {});
         request.response.setHeaders(this.defaultHeaders);
       }
+      else {
+        request = routeHandler.request;
+      }
 
       const e = err instanceof KuzzleError
         ? err
         : errorsManager.getError('request_error', err);
+
       replyWithError(cb, request, e);
+    }
+  }
+
+  /**
+   * Route HTTP messages using an HTTP method that is not handled by Kuzzle's
+   * API, such as OPTIONS.
+   * @param  {HttpMessage} message
+   * @param {function} cb
+   */
+  routeUnhandledHttpMethod(message, cb) {
+    const
+      requestContext = this.kuzzle.router.connections.get(message.connection.id),
+      request = new Request(
+        { requestId: message.requestId },
+        requestContext && requestContext.toJSON());
+
+    request.response.setHeaders(this.defaultHeaders);
+
+    if (message.method === 'OPTIONS') {
+      request.input.headers = message.headers;
+      request.setResult({}, 200);
+
+      this.kuzzle.pipe('http:options', request)
+        .then(result => {
+          cb(result);
+          return null;
+        })
+        .catch(error => replyWithError(cb, request, error));
+
       return;
     }
+
+    replyWithError(
+      cb,
+      request,
+      errorsManager.getError('unrecognized_http_method', message.method));
+  }
+
+  /**
+   * Checks that an incoming HTTP message is well-formed
+   * @param  {HttpMessage} message
+   */
+  validateMessage (message) {
+    const contentType = message.headers['content-type'];
+
+    if (contentType && !contentType.startsWith('application/json')) {
```

Comment:
Nit (don't know if it's worth it here): `String.indexOf` is faster than `String.startsWith`

Reply:
It does.